satellite/repair/repairer/repairer.go

```diff
@@ -38,65 +40,106 @@ type Config struct {
 //
 // architecture: Worker
 type Service struct {
-	log      *zap.Logger
-	queue    queue.RepairQueue
-	config   *Config
-	Limiter  *sync2.Limiter
-	Loop     sync2.Cycle
-	repairer *SegmentRepairer
+	log        *zap.Logger
+	queue      queue.RepairQueue
+	config     *Config
+	JobLimiter *semaphore.Weighted
+	Loop       sync2.Cycle
+	repairer   *SegmentRepairer
 }
 
 // NewService creates repairing service
 func NewService(log *zap.Logger, queue queue.RepairQueue, config *Config, repairer *SegmentRepairer) *Service {
 	return &Service{
-		log:      log,
-		queue:    queue,
-		config:   config,
-		Limiter:  sync2.NewLimiter(config.MaxRepair),
-		Loop:     *sync2.NewCycle(config.Interval),
-		repairer: repairer,
+		log:        log,
+		queue:      queue,
+		config:     config,
+		JobLimiter: semaphore.NewWeighted(int64(config.MaxRepair)),
+		Loop:       *sync2.NewCycle(config.Interval),
+		repairer:   repairer,
 	}
 }
 
 // Close closes resources
 func (service *Service) Close() error { return nil }
 
+// WaitForPendingRepairs waits for all ongoing repairs to complete.
+//
+// NB: this assumes that service.config.MaxRepair will never be changed once this Service instance
+// is initialized. If that is not a valid assumption, we should keep a copy of its initial value to
+// use here instead.
+func (service *Service) WaitForPendingRepairs() {
```

Comment:
This won't work as an exported func. Because at some point there could be a an item in a queue and a worker starting, however the semaphore hasn't been yet acquired. You probably will need a separate `sync2.Fence` to fix this.

Reply:
Similarly, when you call it concurrently with Run in the start, the `WaitForPendingRepairs` could return immediately.

Reply:
This is supposed to have the same semantics as `service.Limiter.Wait()` had. It looks like that also didn't care if there were items in the queue or not. And it is also possible in that case for a worker to have started without yet adding a job with `service.Limiter.Go()`.

Maybe this just needs clearer documentation. What if it said something like

    // WaitForPendingRepairs waits for all ongoing repairs to complete. For the purposes of this
    // function, "ongoing repairs" are specifically repair jobs that have already acquired
    // service.JobLimiter.

Reply:
@egonelbre would that address your concern sufficiently? Or do I not understand the problem yet

Reply:
Sure, comment might be sufficient, but I'm still slightly worried about flakiness.

Reply:
Can you clarify a little about where the flakiness would come in? I hate flakiness too.

Reply:
Yeah, I guess I'm not able to find an exact scenario. I guess it could happen when somebody calls `WaitForPendingRepairs` from outside of tests or `service.Run`